dot/sync/message_test.go

```diff
@@ -0,0 +1,439 @@
+			},
+			args: args{req: &network.BlockRequestMessage{
+				StartingBlock: *variadic.MustNewUint32OrHash(2),
+				Direction:     network.Descending,
+			}},
+			err: nil,
+			want: &network.BlockResponseMessage{BlockData: []*types.BlockData{{
+				Hash: common.Hash{1, 2},
+			}}},
+		},
+		{
+			name: "descending request, endHash not nil",
+			fields: fields{
+				blockState: mockBlockState,
+			},
+			args: args{req: &network.BlockRequestMessage{
+				StartingBlock: *variadic.MustNewUint32OrHash(0),
+				EndBlockHash:  &common.Hash{1, 2, 3},
+				Direction:     network.Descending,
+			}},
+			want: &network.BlockResponseMessage{BlockData: []*types.BlockData{{
+				Hash: common.Hash{1, 2},
+			}}},
+		},
+		{
+			name: "ascending request, startHash, nil endHash",
+			fields: fields{
+				blockState: mockBlockState,
+			},
+			args: args{req: &network.BlockRequestMessage{
+				StartingBlock: *variadic.MustNewUint32OrHash(common.Hash{}),
+				Direction:     network.Ascending,
+			}},
+			want: &network.BlockResponseMessage{BlockData: []*types.BlockData{{
+				Hash: common.Hash{1, 2},
+			}}},
+		},
+		{
+			name: "descending request, startHash, nil endHash",
+			fields: fields{
+				blockState: mockBlockState,
+			},
+			args: args{req: &network.BlockRequestMessage{
+				StartingBlock: *variadic.MustNewUint32OrHash(common.Hash{}),
+				Direction:     network.Descending,
+			}},
+			want: &network.BlockResponseMessage{BlockData: []*types.BlockData{{
+				Hash: common.Hash{1, 2},
+			}}},
+		},
+		{
+			name: "invalid direction",
+			args: args{req: &network.BlockRequestMessage{
+				Direction: network.SyncDirection(3),
+			}},
+			err: errInvalidRequestDirection,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := &Service{
+				blockState:     tt.fields.blockState,
+				chainSync:      tt.fields.chainSync,
+				chainProcessor: tt.fields.chainProcessor,
+				network:        tt.fields.network,
+			}
+			got, err := s.CreateBlockResponse(tt.args.req)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("CreateBlockResponse() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func newMockBlockStateForMessageTest(ctrl *gomock.Controller) BlockState {
+	mock := NewMockBlockState(ctrl)
+
+	mock.EXPECT().GetHashByNumber(gomock.AssignableToTypeOf(uint(0))).DoAndReturn(func(
+		number uint) (common.Hash, error) {
+		return common.Hash{}, nil
+	}).AnyTimes()
+
+	mock.EXPECT().IsDescendantOf(gomock.AssignableToTypeOf(common.Hash{}),
+		gomock.AssignableToTypeOf(common.Hash{})).Return(true, nil).AnyTimes()
+
+	mock.EXPECT().GetHeader(gomock.AssignableToTypeOf(common.Hash{})).DoAndReturn(func(hash common.Hash) (*types.
+		Header, error) {
+		header := &types.Header{
+			Number: uint(hash[0]),
+		}
+		return header, nil
+	}).AnyTimes()
+
+	return mock
+}
```

Comment:
Please remove this helper function (not restrictive enough on mock calls i.e. `AnyTimes()` etc.), and configure the mock in `TestService_checkOrGetDescendantHash` (the only place it's used at)